src/common/io.py

```diff
@@ -2990,6 +2990,163 @@ def get_result(self) -> ResultDymolaBinary:
 
+class _DelayedVarReader4Diags(DelayedVarReader4):
+    def read_sub_array(self, hdr, copy=True):
+        match hdr.name:
+            case b"data_2":
+                return {
+                    "section": "data_2",
+                    "file_position": self.mat_stream.tell(),
+                    "sizeof_type": hdr.dtype.itemsize,
+                    "nbr_points": hdr.dims[1],
+                    "nbr_variables": hdr.dims[0]
+                }
+            case b"data_3":
+                return {
+                    "section": "data_3",
+                    "file_position": self.mat_stream.tell(),
+                    "sizeof_type": hdr.dtype.itemsize,
+                    "nbr_points": hdr.dims[1],
+                    "nbr_variables": hdr.dims[0]
+                }
+            case b"name":
+                return {
+                    "section": "name",
+                    "file_position": self.mat_stream.tell(),
+                    "sizeof_type": hdr.dtype.itemsize,
+                    "max_length": hdr.dims[0],
+                    "nbr_variables": hdr.dims[1]
+                }
+            case _:
+                return super().read_sub_array(hdr, copy)
+
+class _DelayedVariableLoadDiags(DelayedVariableLoad):
+    def initialize_read(self):
+        self.dtypes = convert_dtypes(mdtypes_template, self.byte_order)
+        self._matrix_reader = _DelayedVarReader4Diags(self)
+
+class ResultHandlerCalcDiags(ResultReader):
+    def __init__(self, fname):
+        """
+        Load a .mat result file with diagnostics data explicitly written to data_3.
+
+        Parameters::
+
+            fname --
+                Name of file or a file object supported by scipy.io.loadmat,
+                which the result is written to.
+        """
+
+        if isinstance(fname, str):
+            self._fname = fname
+        elif hasattr(fname, "name") and os.path.isfile(fname.name):
+            self._fname = fname.name
+
+        data_sections = ["name", "dataInfo", "data_2", "data_3"]
+        with open(self._fname, "rb") as f:
+            delayed = _DelayedVariableLoadDiags(f, chars_as_strings=False)
+            self.raw: dict = delayed.get_variables(variable_names = data_sections)
+
+        self._name_info: dict = self.raw["name"]
+        self._dataInfo: dict = self.raw["dataInfo"]
+        self._data_2_info: dict = self.raw["data_2"]
+        self._contains_diagnostic_data: bool = ("data_3" in self.raw)
+        if self._contains_diagnostic_data:
+            self._data_3_info: dict = self.raw["data_3"]
+
+    @cache
+    def _get_variable_name_to_index_dict(self) -> dict[str, int]:
+        name_dict: dict = fmi_util.read_name_list(
+            encode(self._fname),
+            self._name_info["file_position"],
+            int(self._name_info["nbr_variables"]),
+            int(self._name_info["max_length"])
+        )
+
+        return {decode(k): v for k, v in name_dict.items()}
+
+    @cache
```

Comment:
I think we can skip cache for this one as it is just calling another function that is cached. Having all the names 2 times in memory is probably not good as time saved will be negligible.